### PR TITLE
test: happy case with proxy

### PR DESCRIPTION
**Pull request by charleenfei**

### CODE REVIEW

File: bin/test/extract_key.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const { ethers } = require("ethers");
+
+ const key = `{"address":"a49be2371e13049400f0330c26778aae3537087a","crypto":{"cipher":"aes-128-ctr","ciphertext":"c3b92d86a73e856e4eb66470addfa15829e2af9093da04c0555dfb1734d423d0","cipherparams":{"iv":"d5ba72c8dd600b0d3ee9818e041ab3c8"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":262144,"p":1,"r":8,"salt":"d5902caf8e509ea0886f93d008cb99fbd392210c3f7ff24b525df4ce1788a0ac"},"mac":"d92a0cc3fe335f20d5e757804ae2c901a6764eebae49ad2b64139122271fc0a4"},"id":"2c31ff99-ba0e-42aa-bea8-a50d04a8072b","version":3}`
```
Comment (offerijns) [inline bin/test/extract_key.js:3]: Oops... ;-)

File: src/actions/admin.ts [hunk 2]
```diff
@@ -128,6 +128,18 @@ export function AdminActions<ActionsBase extends Constructor<TinlakeParams>>(Bas
...
+
+    getSubmissionPeriod = async () => {
+      return (await this.contract('COORDINATOR').submissionPeriod())
+    }
+
+    getChallengeTime = async () => {
+      return (await this.contract('COORDINATOR').challengeTime()).toBN()
+    }
+
+    getMinimumEpochTime = async () => {
```
Comment (offerijns) [inline src/actions/admin.ts:140]: Ah, I also added this. I think if you pull the `v3` branch, this can be removed.

Comment (offerijns) [reply src/actions/admin.ts:140]: Oh, I added it to `coordinator.ts` though.

File: src/actions/proxy.spec.ts [hunk 3]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
-    //   const amount = '10'
-    //   const relyTx = await governanceTinlake.relyAddress(adminAccount.address, contractAddresses['NFT_FEED'])
-    //   const relyResult = await governanceTinlake.getTransactionReceipt(relyTx)
-
-    //   // console.log('relyResult', relyResult)
-
-    //   const nftfeedId = await adminTinlake.getNftFeedId(testConfig.nftRegistry, Number(tokenId))
-    //   const updateNftTx = await adminTinlake.updateNftFeed(nftfeedId, Number(amount))
-    //   const updateNftResult = await adminTinlake.getTransactionReceipt(updateNftTx)
+  describe('borrow - lend - repay - redeem flow', async () => {
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:38]: 🎉

File: src/actions/proxy.spec.ts [hunk 4]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
 
-    //   // console.log('updateNftResult', updateNftResult)
+      await governanceTinlake.setMinimumJuniorRatio('0')
+      const ratio = await governanceTinlake.getMaxJuniorRatio()
+      assert.equal(ratio, 1000000000000000000000000000)
 
-    //   await fundTranche('100')
-    //   const initialTrancheBalance = await borrowerTinlake.getCurrencyBalance(contractAddresses['JUNIOR_TRANCHE'])
+      await governanceTinlake.setMinimumEpochTime('1')
+      const epoch = await governanceTinlake.getMinimumEpochTime()
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:46]:
```suggestion
      const minimumEpochTime = await governanceTinlake.getMinimumEpochTime()
```

File: src/actions/proxy.spec.ts [hunk 5]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
+      assert.equal(epoch.toString(), '1')
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:47]:
```suggestion
      assert.equal(minimumEpochTime.toString(), '1')
```

File: src/actions/proxy.spec.ts [hunk 6]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
-//   // mint currency for lender
-//   const mintTx = await governanceTinlake.mintCurrency(lenderAccount.address, amount)
-//   governanceTinlake.getTransactionReceipt(mintTx)
-
-//   // lender supplies tranche with funds
-//   const supplyTx = await lenderTinlake.supplyJunior(amount)
-//   await lenderTinlake.getTransactionReceipt(supplyTx)
-// }
+// TODO: move to utils, requires moving xTinlake
+async function verifyTransactionSuccess(transaction: any) {
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:126]:
```suggestion
async function verifyTransactionSuccess(transaction: PendingTransaction) {
```

File: src/actions/proxy.spec.ts [hunk 7]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
+  const transactionResult = await lenderTinlake.getTransactionReceipt(transaction)
+  assert.equal(transactionResult.status, SUCCESS_STATUS)
+}
+
+async function redeemTranche(amount: string, tranche: string){
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:131]:
```suggestion
async function redeemTranche(amount: string, tranche: 'senior' | 'junior'){
```

File: src/actions/proxy.spec.ts [hunk 8]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
+    await verifyTransactionSuccess(solveTx)
+
+    const srCalc2 = await governanceTinlake.calcSeniorDisburse(lenderAccount.address)
+    assert.equal(srCalc2.payoutTokenAmount, 0)
+    const seniorTokenBalance2 = await governanceTinlake.getSeniorTokenBalance(lenderAccount.address)
+    assert.equal(seniorTokenBalance2, 0)
+  }
+}
+
+async function fundTranche(amount: string, tranche: string) {
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:183]:
```suggestion
async function fundTranche(amount: string, tranche: 'senior' | 'junior') {
```

File: src/actions/proxy.spec.ts [hunk 9]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
+      // add governance address to memberlists
+      await governanceTinlake.updateJuniorMemberList(contractAddresses['GOVERNANCE'], validityTimestamp)
+      await governanceTinlake.updateSeniorMemberList(contractAddresses['GOVERNANCE'], validityTimestamp)
 
-    //   // console.log('borrowResult', borrowResult)
+      // create new proxy and mint collateral NFT to borrower
+      const proxyAddr = await borrowerTinlake.proxyCreateNew(borrowerAccount.address)
+      const tokenId = await governanceTinlake.mintTitleNFT(testConfig.nftRegistry, borrowerAccount.address)
+      const approveTx = await borrowerTinlake.approveNFT(testConfig.nftRegistry, tokenId, proxyAddr)
+      await verifyTransactionSuccess(approveTx)
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:61]: Great idea to add this helper! Definitely makes the intent of the tests much clearer.

File: src/actions/proxy.spec.ts [hunk 10]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
-    //   const initialTrancheBalance = await borrowerTinlake.getCurrencyBalance(contractAddresses['JUNIOR_TRANCHE'])
+      await governanceTinlake.setMinimumEpochTime('1')
+      const epoch = await governanceTinlake.getMinimumEpochTime()
+      assert.equal(epoch.toString(), '1')
 
-    //   // borrow
-    //   const borrowTx = await borrowerTinlake.proxyLockBorrowWithdraw(proxyAddr, loanId, amount, borrowerAccount.address)
-    //   // console.log('borrowTx', borrowTx)
+      await governanceTinlake.setMinimumChallengeTime('1')
+      const challenge = await governanceTinlake.getChallengeTime()
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:50]:
```suggestion
      const challengeTime = await governanceTinlake.getChallengeTime()
```

File: src/actions/proxy.spec.ts [hunk 11]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
+      assert.equal(challenge.toString(), '1')
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:51]:
```suggestion
      assert.equal(challengeTime.toString(), '1')
```

File: src/actions/proxy.spec.ts [hunk 12]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
 const { SUCCESS_STATUS, FAIL_STATUS, FAUCET_AMOUNT, contractAddresses } = testConfig
 
+const SENIOR = 'senior'
+const JUNIOR = 'junior'
+
+const juniorAmount = '300'
+const seniorAmount = '100'
+const amount = '100';
+
+// valid until Sunday, 21 December 2025
```
Comment (offerijns) [inline src/actions/proxy.spec.ts:23]: Better to be safe than sorry 🤣

**Review by offerijns — APPROVED** [review]
Great work!

File: src/actions/admin.ts (hunk 2 above)
Comment (charleenfei) [reply src/actions/admin.ts:140]: i can rm it from admin.ts

File: src/actions/proxy.spec.ts [hunk 13]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
-    //   const borrowTx = await borrowerTinlake.proxyLockBorrowWithdraw(proxyAddr, loanId, amount, borrowerAccount.address)
-    //   // console.log('borrowTx', borrowTx)
+      await governanceTinlake.setMinimumChallengeTime('1')
+      const challengeTime = await governanceTinlake.getChallengeTime()
+      assert.equal(challengeTime.toString(), '1')
 
-    //   const borrowResult = await borrowerTinlake.getTransactionReceipt(borrowTx)
+      // add governance address to memberlists
+      await governanceTinlake.updateJuniorMemberList(contractAddresses['GOVERNANCE'], validityTimestamp)
+      await governanceTinlake.updateSeniorMemberList(contractAddresses['GOVERNANCE'], validityTimestamp)
```
Comment (philipstanislaus) [inline src/actions/proxy.spec.ts:55]: Should we not add `lenderAccount` here?

File: src/actions/proxy.spec.ts [hunk 14]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
+      const borrowTx = await borrowerTinlake.proxyBorrowWithdraw(proxyAddr, loanId, amount, borrowerAccount.address);
+      await verifyTransactionSuccess(borrowTx)
 
-    // it('fail: does not succeed if the proxy is not approved to take the NFT', async () => {
-    //   const proxyAddr = await borrowerTinlake.proxyCreateNew(borrowerAccount.address)
-    //   const tokenId = `${Math.floor(Math.random() * 10e15) + 1}`
+      const borrowerValue = await governanceTinlake.getCurrencyBalance(borrowerAccount.address)
+      const newReserveValue = await governanceTinlake.getCurrencyBalance(contractAddresses['RESERVE'])
+      // assert.equal(borrowerValue.toString(), principal.toString())
+      // assert.equal(newReserveValue.toString(), initialReserveValue.sub(principal).toString());
```
Comment (philipstanislaus) [inline src/actions/proxy.spec.ts:99]: Why are those commented out?

**Review by philipstanislaus — APPROVED** [review]
This is awesome, thanks @charleenfei ! 🎉

File: src/actions/proxy.spec.ts [hunk 15]
```diff
@@ -3,204 +3,216 @@ import { createTinlake, TestProvider } from '../test/utils'
...
-
-    //   const nftfeedId = await adminTinlake.getNftFeedId(testConfig.nftRegistry, Number(tokenId))
-    //   const updateNftTx = await adminTinlake.updateNftFeed(nftfeedId, Number(amount))
-    //   const updateNftResult = await adminTinlake.getTransactionReceipt(updateNftTx)
+  describe('borrow - lend - repay - redeem flow', async () => {
+    it('success: full loan cycle', async () => {
 
-    //   // console.log('updateNftResult', updateNftResult)
+      await governanceTinlake.setMinimumJuniorRatio('0')
+      const ratio = await governanceTinlake.getMaxJuniorRatio()
```
[the remark this answers is not in the document]
Comment (charleenfei) [reply src/actions/proxy.spec.ts:42]: I can also use getMinJuniorRatio but in principle it should be the same right? 0 -> 100

File: src/actions/proxy.spec.ts (hunk 13 above)
Comment (charleenfei) [reply src/actions/proxy.spec.ts:55]: we add it in the fundTranches function

**Pull request merged**